ext/bcmath/libbcmath/src/round.c

```diff
@@ -0,0 +1,172 @@
+/*
+   +----------------------------------------------------------------------+
+   | Copyright (c) The PHP Group                                          |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | https://www.php.net/license/3_01.txt                                 |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+   | Authors: Saki Takamachi <saki@php.net>                               |
+   +----------------------------------------------------------------------+
+*/
+
+#include "bcmath.h"
+#include "private.h"
+#include <stddef.h>
+
+void bc_round(bc_num num, zend_long precision, zend_long mode, bc_num *result)
+{
+	/* clear result */
+	bc_free_num(result);
+
+	/*
+	* The following cases result in an early return:
+	*
+	* - When rounding to an integer part which is larger than the number
+	*   e.g. Rounding 21.123 to 3 digits before the decimal point.
+	* - When rounding to a greater decimal precision then the number has, the number is unchanged
+	*   e.g. Rounding 21.123 to 4 digits after the decimal point.
+	* - If the fractional part ends with zeros, the zeros are omitted and the number of digits in num is reduced.
+	*   Meaning we might end up in the previous case.
+	*/
+	if (precision < 0 && num->n_len < (size_t) (-(precision + Z_L(1))) + 1) {
+		*result = bc_copy_num(BCG(_zero_));
+		return;
+	}
+	if (precision >= 0 && num->n_scale <= precision) {
+		*result = bc_copy_num(num);
+		return;
+	}
+
+	/*
+	 * If the calculation result is a negative value, there is an early return,
+	 * so no underflow will occur.
+	 */
+	size_t rounded_len = num->n_len + precision;
+
+	/*
+	 * Initialize result
+	 * For example, if rounded_len is 0, it means trying to round 50 to 100 or 0.
+	 * If the result of rounding is carried over, it will be added later, so first set it to 0 here.
+	 */
+	if (rounded_len == 0) {
+		*result = bc_copy_num(BCG(_zero_));
+	} else {
+		*result = bc_new_num(num->n_len, precision > 0 ? precision : 0);
+		memcpy((*result)->n_value, num->n_value, rounded_len);
+	}
+	(*result)->n_sign = num->n_sign;
+
+	const char *nptr = num->n_value + rounded_len;
+
+	/* Check cases that can be determined without looping. */
+	switch (mode) {
+		case PHP_ROUND_HALF_UP:
+			if (*nptr >= 5) {
+				goto up;
+			} else if (*nptr < 5) {
+				return;
+			}
+			break;
+
+		case PHP_ROUND_HALF_DOWN:
```

Comment:
Can't the half down case be determined without a loop? If it does require a loop how come HALF UP doesn't need one?

Reply:
For example, let's compare two cases: `5.0000` and `5.0001`.

HALF_UP: Either way, it'll be a 10.
HALF_DOWN: 5.0001 will be 10, but 5.0000 will be 0.

Therefore, in the case of HALF_UP, there is no need to loop because it always carries up regardless of the loop result.

Reply:
Oh wait, I was refactoring BCMath in another PR and found that trailing zeros are discarded when converting strings to bc_num structs.

So in all cases I don't need to loop, just check if there is one more digit.